plugins/lcvcore/src/qgradient.cpp

```diff
@@ -0,0 +1,124 @@
+#include "qgradient.h"
+#include <QVector3D>
+#include "qmat.h"
+#include "opencv2/core.hpp"
+#include <opencv2/core/matx.hpp>
+
+#define GAMMA 0.43
+
+QGradient::QGradient(QObject *parent) : QObject(parent)
+{
+
+}
+
+void QGradient::draw(QMat* result, QPointF p1, QPointF p2, QColor c1, QColor c2)
+{
+    if (p1 == p2) return;
+
+    auto mat = result->cvMat();
+    if (mat->rows == 0 || mat->cols == 0) return;
+
+
+    start = p1; end = p2;
+    startColor = c1; endColor = c2;
+
+    calculateLineParams();
+
+    linearSC = linearFromRGB(startColor);
+    linearEC = linearFromRGB(endColor);
+    brightness1 = pow(linearSC.x()+linearSC.y()+linearSC.z(), GAMMA);
+    brightness2 = pow(linearEC.x()+linearEC.y()+linearEC.z(), GAMMA);
+
+    for (int y = 0; y < mat->rows; ++y)
+    for (int x = 0; x < mat->cols; ++x)
+    {
+        cv::Vec3b& color = mat->at<cv::Vec3b>(cv::Point(x,y));

```

Comment:
The "cv::Mat::at" method in open cv is quite slow when accessing pixel values. Direct pointer access is a few times faster:

```cpp
for(int y = 0; y < mat->rows; y++)
{
    uchar* p = mat->ptr<uchar>(y); // pointer to current row
   
    for(int x = 0; x < mat->cols; x++)
        uchar b = p[x * mat->channels + 0]
        uchar g = p[x * mat->channels + 1]
        uchar r = p[x * mat->channels + 2]
}
```